internal/instance/resource_instance.go

```diff
@@ -888,6 +971,66 @@ func (r InstanceResource) createInstanceFromImage(ctx context.Context, server in
 	return diags
 }
 
+func (r InstanceResource) createInstanceFromSourceFile(ctx context.Context, server incus.InstanceServer, plan InstanceModel) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	name := plan.Name.ValueString()
+
+	var poolName string
+
+	if len(plan.Devices.Elements()) > 0 {
+		// Only one device is expected, this is ensured by ValidateConfig.
+		deviceList := make([]common.DeviceModel, 0, 1)
+		diags = plan.Devices.ElementsAs(ctx, &deviceList, false)
+		if diags.HasError() {
+			return diags
+		}
+
+		// Exactly one property named "pool" is expected, this is ensured by ValidateConfig.
+		properties := make(map[string]string, 1)
+		diags = deviceList[0].Properties.ElementsAs(ctx, &properties, false)
+		if diags.HasError() {
+			return diags
+		}
+
+		poolName = properties["pool"]
```

Comment:
Won't that still fail if a user passes us a device that's of type `disk` and has both:
 - path=/
 - pool=default

As would normally be required for a root device to be valid?

Reply:
Oh, I see, this was then handled in a follow-up commit.

Can you rebase that last commit into the 3rd one? Makes reviewing commits individually easier :)